pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/PubsubProducer.java

```diff
@@ -0,0 +1,401 @@
+        this.keySerializer =
+            configs.getConfiguredInstance(
+                PubsubProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.keySerializer.configure(configs.originals(), true);
+      } else {
+        configs.ignore(PubsubProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+        this.keySerializer = keySerializer;
+      }
+
+      if (valueSerializer == null) {
+        this.valueSerializer = configs.getConfiguredInstance(
+            PubsubProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.valueSerializer.configure(configs.originals(), false);
+      } else {
+        configs.ignore(PubsubProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+        this.valueSerializer = valueSerializer;
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    batchSize = configs.getInt(PubsubProducerConfig.BATCH_SIZE_CONFIG);
+    isAcks = configs.getString(PubsubProducerConfig.ACKS_CONFIG).matches("1|all");
+    project = configs.getString(PubsubProducerConfig.PROJECT_CONFIG);
+    maxRequestSize = configs.getInt(PubsubProducerConfig.MAX_REQUEST_SIZE_CONFIG);
+    perTopicBatches = Collections.synchronizedMap(new HashMap<>());
+
+    log.debug("Producer successfully initialized.");
+  }
+
+  /**
+   * Send the given record asynchronously and return a future which will eventually contain the response information.
+   *
+   * @param record The record to send
+   * @return A future which will eventually contain the response information
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+    return send(record, null);
+  }
+
+  /**
+   * Send a record and invoke the given callback when the record has been acknowledged by the server
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+    log.trace("Received " + record.toString());
+    if (closed) {
+      throw new RuntimeException("Publisher is closed");
+    }
+
+    String topic = record.topic();
+    Map<String, String> attributes = new HashMap<>();
+
+    byte[] serializedKey = ByteString.EMPTY.toByteArray();
+    if (record.key() != null) {
+      serializedKey = this.keySerializer.serialize(topic, record.key());
+      attributes.put(channelUtil.KEY_ATTRIBUTE, new String(serializedKey, StandardCharsets.ISO_8859_1));
+    }
+
+    if (project == null) {
+      throw new RuntimeException("No project specified.");
+    }
+
+    byte[] valueBytes = ByteString.EMPTY.toByteArray();
+    if (record.value() != null) {
+      valueBytes = valueSerializer.serialize(topic, record.value());
+    }
+
+    checkRecordSize(Records.LOG_OVERHEAD + Record.recordSize(serializedKey, valueBytes));
+
+    PubsubMessage message =
+        PubsubMessage.newBuilder()
+          .setData(ByteString.copyFrom(valueBytes))
+          .putAllAttributes(attributes)
+          .build();
+    List<PubsubMessage> batch = perTopicBatches.get(topic);
```

Comment:
You'll need some kind of mutual exclusion to ensure you don't have concurrency issues with simultaneous calls to send.